File: minimum-limit-of-balls-in-a-bag/minimum-limit-of-balls-in-a-bag.cpp
class Solution {
public:
    int minimumSize(vector<int>& nums, int maxOperations) {
        int start=1,end=1e9;
        while(start<end){
            int mid=start+(end-start)/2;
            if(isValid(mid,nums,maxOperations)){
                end=mid;
            }else{
                start=mid+1;
            }
        }
        
        return start;
    }
    
    bool isValid(int x,vector<int>& nums,int maxOperations){
        int total=0;
        for(auto num : nums){
            total+=(num-1)/x;
        }
        
        return total<=maxOperations;
    }
};

File: minimum-limit-of-balls-in-a-bag/minimum-limit-of-balls-in-a-bag.java
class Solution {
    public int minimumSize(int[] nums, int maxOperations) {
        int start = 1,end = 1000000000;
        while(start < end){
            int mid = start + (end - start) / 2;
            if(isValid(mid,nums,maxOperations)){
                end = mid;
            }else{
                start = mid + 1;
            }
        }
        
        return start;
    }
    
    public boolean isValid(int mid,int nums[],int maxOperations){
        int total = 0;
        for(int num : nums){
            total += (num - 1)/mid;
        }
        
        return total <= maxOperations;
    }
}